File: go/0141_linked_list_cycle.go
package main

// type ListNode struct {
// 	Val  int
// 	Next *ListNode
// }

// time: O(n)
// space O(1)

func hasCycle(head *ListNode) bool {
	s := head
	f := head
	for f != nil && f.Next != nil {
		s = s.Next
		f = f.Next.Next
		if s == f {
			return true
		}
	}

	return false
}

File: typescript/solutions/0141-linked-list-cycle.ts
import { ListNode } from "./0206-reverse-linked-list"

// time: O(n)
// space: O(1)

function hasCycle(head: ListNode | null): boolean {
  let slow = head
  let fast = head

  while (fast?.next?.next) {
    if (slow === fast) return true
    slow = slow?.next!
    fast = fast?.next.next
  }

  return false
}
